src/main/host/descriptor/tcp.c

```diff
@@ -2379,13 +2378,19 @@ static gssize _tcp_sendUserData(LegacySocket* socket, const Thread* thread, Plug
     gsize maxPacketLength = CONFIG_TCP_MAX_SEGMENT_SIZE;
     gsize bytesCopied = 0;
 
+    if (buffer.val == 0) {
```

Comment:
Shouldn't need to explicitly check for NULL; `MemoryManager` will return an `EFAULT` in the usual case that 0x0 isn't accessible. I think it's technically possible to map a page at address 0, in which case an explicit NULL check would incorrectly return an error. NULL might also be ok if `nBytes` is 0.

As a general programming rule it's less duplicative and more reliable to try an operation and see if it succeeds than to try to figure out if an operation will succeed before trying it.

Reply:
I think I'd rather leave it the same as the original C code. I tried bubbling up the error from the memory manager, but then you also need to undo all of the TCP state changes that happen when creating the packet object. For example creating the packet object sometimes increases the TCP object's sequence number. So if you fail to set the packet's payload, then you need to decrease the sequence number back. But there are only certain conditions where the sequence number is increased, and so you need to copy that same logic to decide whether you need to decrease the sequence number. Since this code is fragile, I'd rather leave the control flow to be the same. This `buffer.val == 0` used to be in the C syscall handler, but I moved it here to the TCP code so that it wouldn't be used for all other socket types as well.

Reply:
Right, previously Shadow would segfault if you give an invalid non-NULL plugin pointer. With this PR it will panic instead. I'll add a comment, but I'm not sure it's worth fixing for the C TCP code.

Reply:
Ok, sounds good :)